File: frontend/src/types/ai.ts
export interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

export interface Point2D {
  x: number;
  y: number;
}

export interface TrackedObject {
  track_id: number;
  camera_id: string;
  object_type: string;
  category: 'person' | 'vehicle' | 'animal' | 'drone' | 'other';
  confidence: number;
  bbox: BoundingBox;
  center: Point2D;
  bottom_center: Point2D;
  first_seen_at: string;
  last_seen_at: string;
  frame_count: number;
  speed: number;
  direction: 'NORTH' | 'SOUTH' | 'EAST' | 'WEST' | 'NORTH_EAST' | 'NORTH_WEST' | 'SOUTH_EAST' | 'SOUTH_WEST' | 'STATIONARY' | 'UNKNOWN';
  tracking_state: 'DETECTED' | 'TRACKING' | 'TEMPORARILY_LOST' | 'REACQUIRED' | 'EXPIRED';
  trajectory: Point2D[];
}

export interface CameraAICounters {
  people: number;
  vehicles: number;
  animals: number;
  other: number;
  total_tracks: number;
}

export interface CameraAIStatus {
  camera_id: string;
  status: 'ACTIVE' | 'STARTING' | 'PAUSED' | 'ERROR' | 'NO_STREAM';
  inference_fps: number;
  latency_ms: number;
  frames_processed: number;
  frames_skipped: number;
  active_tracks_count: number;
  counters: CameraAICounters;
  device: string;
  model_name: string;
  error_message?: string | null;
}

export interface CameraAIConfig {
  camera_id: string;
  enabled: boolean;
  model_name: string;
  target_fps: number;
  input_size: number;
  device: string;
  conf_person: number;
  conf_vehicle: number;
  conf_animal: number;
  conf_drone: number;
  conf_other: number;
  track_thresh: number;
  match_thresh: number;
  max_lost_frames: number;
  max_trajectory_length: number;
}

export interface CameraAIConfigUpdate {
  enabled?: boolean;
  model_name?: string;
  target_fps?: number;
  conf_person?: number;
  conf_vehicle?: number;
  conf_animal?: number;
  conf_drone?: number;
  conf_other?: number;
  track_thresh?: number;
  match_thresh?: number;
}

export interface AIRealTimeTelemetryMessage {
  event: string;
  camera_id: string;
  timestamp: string;
  status: 'ACTIVE' | 'STARTING' | 'PAUSED' | 'ERROR' | 'NO_STREAM';
  inference_fps: number;
  latency_ms: number;
  counters: CameraAICounters;
  tracks: TrackedObject[];
}

export interface AIMetrics {
  active_workers: number;
  total_cameras_configured: number;
  total_frames_processed: number;
  avg_latency_ms: number;
  total_active_tracks: number;
  model: string;
  device: string;
}

File: backend/app/schemas/ai.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime

class BoundingBox(BaseModel):
    x: float = Field(..., description="Top-left x coordinate in pixels")
    y: float = Field(..., description="Top-left y coordinate in pixels")
    width: float = Field(..., description="Box width in pixels")
    height: float = Field(..., description="Box height in pixels")

class Point2D(BaseModel):
    x: float
    y: float

class DetectionResult(BaseModel):
    class_name: str
    category: str               # person, vehicle, animal, drone, other
    confidence: float
    bbox: BoundingBox
    timestamp: datetime
    camera_id: str

class TrackedObject(BaseModel):
    track_id: int
    camera_id: str
    object_type: str
    category: str               # person, vehicle, animal, drone, other
    confidence: float
    bbox: BoundingBox
    center: Point2D
    bottom_center: Point2D
    first_seen_at: datetime
    last_seen_at: datetime
    frame_count: int
    speed: float = 0.0          # Relative image-space movement speed (px/sec)
    direction: str = "UNKNOWN"  # NORTH, SOUTH, EAST, WEST, NORTH_EAST, NORTH_WEST, SOUTH_EAST, SOUTH_WEST, STATIONARY, UNKNOWN
    tracking_state: str = "TRACKING" # DETECTED, TRACKING, TEMPORARILY_LOST, REACQUIRED, EXPIRED
    trajectory: List[Point2D] = []

class CameraAICounters(BaseModel):
    people: int = 0
    vehicles: int = 0
    animals: int = 0
    other: int = 0
    total_tracks: int = 0

class CameraAIStatus(BaseModel):
    camera_id: str
    status: str                 # ACTIVE, STARTING, PAUSED, ERROR, NO_STREAM
    inference_fps: float = 0.0
    latency_ms: float = 0.0
    frames_processed: int = 0
    frames_skipped: int = 0
    active_tracks_count: int = 0
    counters: CameraAICounters
    device: str = "CPU"
    model_name: str = "yolov8n"
    error_message: Optional[str] = None

class CameraAIConfigSchema(BaseModel):
    camera_id: str
    enabled: bool
    model_name: str
    target_fps: float
    input_size: int
    device: str
    conf_person: float
    conf_vehicle: float
    conf_animal: float
    conf_drone: float
    conf_other: float
    track_thresh: float
    match_thresh: float
    max_lost_frames: int
    max_trajectory_length: int

    class Config:
        from_attributes = True

class CameraAIConfigUpdate(BaseModel):
    enabled: Optional[bool] = None
    model_name: Optional[str] = None
    target_fps: Optional[float] = None
    conf_person: Optional[float] = None
    conf_vehicle: Optional[float] = None
    conf_animal: Optional[float] = None
    conf_drone: Optional[float] = None
    conf_other: Optional[float] = None
    track_thresh: Optional[float] = None
    match_thresh: Optional[float] = None

class AIRealTimeTelemetryMessage(BaseModel):
    event: str = "AI_DETECTION_TELEMETRY"
    camera_id: str
    timestamp: str
    status: str
    inference_fps: float
    latency_ms: float
    counters: CameraAICounters
    tracks: List[TrackedObject]

class ModelStatusItem(BaseModel):
    model_name: str
    model_path: Optional[str] = None
    configured: bool
    file_exists: bool
    loaded: bool
    status: str  # NOT_CONFIGURED, FILE_MISSING, LOADING, LOADED, ERROR
    error: Optional[str] = None
    capabilities: List[str] = []

class AIModelsOverviewResponse(BaseModel):
    models: List[ModelStatusItem]
    system_ready: bool
